src/inc/til/rle.h

```diff
@@ -0,0 +1,1092 @@
+            // 2. Fix mybase to match
+            // 3. value_type needs to be whatever makes sense to come off of *iterator
+            // 4. difference_type needs to come from somewhere else, probably.
+
+            using mybase = rle_const_iterator<ParentIt>;
+
+            using iterator_category = std::random_access_iterator_tag;
+            using value_type = typename ParentIt::value_type::first_type;
+            using pointer = value_type*;
+            using reference = value_type&;
+            using difference_type = typename ParentIt::difference_type;
+
+            // Use base's constructor.
+            using mybase::mybase;
+
+            [[nodiscard]] reference operator*() const noexcept
+            {
+                return const_cast<reference>(mybase::operator*());
+            }
+
+            [[nodiscard]] pointer operator->() const noexcept
+            {
+                return const_cast<std::remove_const_t<value_type>*>(mybase::operator->());
+            }
+
+            rle_iterator& operator++() noexcept
+            {
+                mybase::operator++();
+                return *this;
+            }
+
+            rle_iterator operator++(int) noexcept
+            {
+                rle_iterator tmp = *this;
+                mybase::operator++();
+                return tmp;
+            }
+
+            rle_iterator& operator--() noexcept
+            {
+                mybase::operator--();
+                return *this;
+            }
+
+            rle_iterator operator--(int) noexcept
+            {
+                rle_iterator tmp = *this;
+                mybase::operator--();
+                return tmp;
+            }
+
+            rle_iterator& operator+=(const difference_type offset) noexcept
+            {
+                mybase::operator+=(offset);
+                return *this;
+            }
+
+            [[nodiscard]] rle_iterator operator+(const difference_type offset) const noexcept
+            {
+                rle_iterator tmp = *this;
+                return tmp += offset;
+            }
+
+            rle_iterator& operator-=(const difference_type offset) noexcept
+            {
+                mybase::operator-=(offset);
+                return *this;
+            }
+
+            // Use base's difference method.
+            using mybase::operator-;
+
+            [[nodiscard]] rle_iterator operator-(const difference_type offset) const noexcept
+            {
+                rle_iterator tmp = *this;
+                return tmp -= offset;
+            }
+
+            [[nodiscard]] reference operator[](const difference_type offset) const noexcept
+            {
+                return const_cast<reference>(mybase::operator[](offset));
+            }
+        };
+    };
+
+    template<typename T, typename S = size_t>
+    class rle
+    {
+    private:
+        boost::container::small_vector<std::pair<T, S>, 1> _list;

```

Comment:
![image](https://user-images.githubusercontent.com/18221333/112511543-beebf680-8d4f-11eb-8371-e33d8a9d7514.png)

@DHowett comment placed in appropriate spot :P